Userland/test.cpp

```diff
@@ -0,0 +1,524 @@
+            return make<NumericCompare>(lhs, rhs, NumericCompare::Equal);
+        } else if (arg == "-ge") {
+            StringView rhs = argv[++optind];
+            return make<NumericCompare>(lhs, rhs, NumericCompare::GreaterOrEqual);
+        } else if (arg == "-gt") {
+            StringView rhs = argv[++optind];
+            return make<NumericCompare>(lhs, rhs, NumericCompare::Greater);
+        } else if (arg == "-le") {
+            StringView rhs = argv[++optind];
+            return make<NumericCompare>(lhs, rhs, NumericCompare::LessOrEqual);
+        } else if (arg == "-lt") {
+            StringView rhs = argv[++optind];
+            return make<NumericCompare>(lhs, rhs, NumericCompare::Less);
+        } else if (arg == "-ne") {
+            StringView rhs = argv[++optind];
+            return make<NumericCompare>(lhs, rhs, NumericCompare::NotEqual);
+        } else if (arg == "-ef") {
+            StringView rhs = argv[++optind];
+            return make<FileCompare>(lhs, rhs, FileCompare::Same);
+        } else if (arg == "-nt") {
+            StringView rhs = argv[++optind];
+            return make<FileCompare>(lhs, rhs, FileCompare::ModificationTimestampGreater);
+        } else if (arg == "-ot") {
+            StringView rhs = argv[++optind];
+            return make<FileCompare>(lhs, rhs, FileCompare::ModificationTimestampLess);
+        } else if (arg.is_null()) {
+            return make<StringCompare>("", lhs, StringCompare::NotEqual);
+        }
+    }
+
+    fatal_error("Unsupported option \033[1m%s", argv[optind]);
+}
+
+OwnPtr<Condition> parse_complex_expression(char* argv[])
+{
+    auto command = parse_simple_expression(argv);
+
+    while (command && argv[optind] && argv[optind + 1]) {
+        StringView arg = argv[++optind];
+
+        enum {
+            AndOp,
+            OrOp,
+        } binary_operation { AndOp };
+
+        if (arg == "-a") {
+            optind++;
+            binary_operation = AndOp;
+        } else if (arg == "-o") {
+            optind++;
+            binary_operation = OrOp;
+        } else {
+            // Ooops, looked too far.
+            optind--;
+            return command;
+        }
+        auto rhs = parse_complex_expression(argv);
+        if (!rhs)
+            fatal_error("Missing right-hand side");
+
+        if (binary_operation == AndOp)
+            command = make<And>(command.release_nonnull(), rhs.release_nonnull());
+        else
+            command = make<Or>(command.release_nonnull(), rhs.release_nonnull());
+    }
+
+    if (g_was_invoked_as_bracket && StringView { argv[optind] } != "]")
+        fatal_error("test invoked as '[' requires a closing bracket ']'");
+
+    return command;
+}
+
+int main(int, char* argv[])
+{
+    if (pledge("stdio rpath", nullptr) < 0) {
+        perror("pledge");
+        return 126;
+    }
+
+    if (LexicalPath { argv[0] }.basename() == "[")
+        g_was_invoked_as_bracket = true;
```

Comment:
Instead of using this flag, how about we do this:
```suggestion
    if (LexicalPath { argv[0] }.basename() == "[") {
        if (StringView { argv[argc - 1] } != "]")
            fatal_error("test invoked as '[' requires a closing bracket ']'");
        argv[argc - 1][0] = '\0';
    }
```
and let the program parse the arguments as if it was invoked as `test`?

Reply:
I was thinking of this solution for the `[ -n ] ]` thing @alimpfard .

Reply:
With this method we also don't have to care about `g_was_invoked_as_bracket`

Reply:
That's...actually pretty nice, yeah.